### PR TITLE
KAFKA-12967; KRaft broker should forward DescribeQuorum to controller

### PR DESCRIPTION
**Pull request by hachikuji**
We added the DescribeQuorum API in KIP-595. This patch adds the logic to forward DescribeQuorum requests to the controller when KRaft is enabled.

Note that the KRaft broker listener has already been enabled in DescribeQuorumRequest.json. The zk broker is not enabled, however, so DescribeQuorum requests will not be advertised and will be rejected at the network layer.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 1]
```diff
@@ -462,16 +461,55 @@ class KafkaApisTest {
...
+    val requestData = DescribeQuorumRequest.singletonRequest(KafkaRaftServer.MetadataPartition)
+    val requestBuilder = new DescribeQuorumRequest.Builder(requestData)
+    metadataCache = MetadataCache.raftMetadataCache(brokerId)
+
+    testForwardableApi(
+      createKafkaApis(raftSupport = true),
+      ApiKeys.DESCRIBE_QUORUM,
+      requestBuilder
+    )
+  }
```
Comment (jsancio) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:492]: Sounds good. How do we check that the `ControlerApi` is doing the right thing :tm: ? Should we add a test to `RaftClusterTest` or do we need to add this to the `AdminClient` first to do that?

**Review by jsancio — COMMENTED** [review]
Thanks for the changes @hachikuji . I just have one question.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:492]: Fair question. Let me see if I can add a request test.

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:492]: Added a basic integration test case. Good thing too because it uncovered the need to set the `forwardable` flag in `ApiKeys`.

**Review by jsancio — APPROVED** [review]
LGTM. Thanks for the fix.

File: core/src/test/java/kafka/test/annotation/Type.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
  * The type of cluster config being requested. Used by {@link kafka.test.ClusterConfig} and the test annotations.
  */
 public enum Type {
-    RAFT {
+    KRAFT {
```
Comment (abbccdda) [inline core/src/test/java/kafka/test/annotation/Type.java:31]: Is this change needed?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 3]
```diff
@@ -462,16 +461,55 @@ class KafkaApisTest {
...
+    apiKey: ApiKeys,
+    requestBuilder: AbstractRequest.Builder[_ <: AbstractRequest],
+  ): Unit = {
     val topicHeader = new RequestHeader(apiKey, apiKey.latestVersion,
       clientId, 0)
 
     val request = buildRequest(requestBuilder.build(topicHeader.apiVersion))
 
-    EasyMock.expect(controller.isActive).andReturn(false)
+    EasyMock.expect(controller.isActive).andStubReturn(false)
```
Comment (abbccdda) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: What's the purpose of this switch?

File: core/src/test/java/kafka/test/annotation/Type.java (hunk 2 above)
Comment (hachikuji) [reply core/src/test/java/kafka/test/annotation/Type.java:31]: Not needed for correctness of course, but KRaft seems to be the preferred term these days for the Kafka+Raft clusters. I thought I would go ahead and change here since there are only a few uses so far.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: I was being a little lazy. For `DescribeQuorum` in KRaft clusters, we forward it to the controller blindly because the broker cannot be a controller. Let me see if I can make this conditional.

File: core/src/test/java/kafka/test/annotation/Type.java (hunk 2 above)
Comment (mumrah) [reply core/src/test/java/kafka/test/annotation/Type.java:31]: Seems fine to change this here. I've been meaning to do this anyways.

File: core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.kafka.common.requests.DescribeQuorumRequest.singletonRequest
+import org.apache.kafka.common.requests.{DescribeQuorumRequest, DescribeQuorumResponse}
+import org.junit.jupiter.api.Assertions._
+import org.junit.jupiter.api.extension.ExtendWith
+
+import scala.jdk.CollectionConverters._
+
+@ExtendWith(value = Array(classOf[ClusterTestExtensions]))
+@ClusterTestDefaults(clusterType = Type.KRAFT)
+class DescribeQuorumRequestTest(cluster: ClusterInstance) {
```
Comment (mumrah) [inline core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:36]: Is it worth adding a test for ZK mode here? Just to see that it fails

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -217,6 +227,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ApiKeys.DESCRIBE_TRANSACTIONS => handleDescribeTransactionsRequest(request)
         case ApiKeys.LIST_TRANSACTIONS => handleListTransactionsRequest(request)
         case ApiKeys.ALLOCATE_PRODUCER_IDS => maybeForwardToController(request, handleAllocateProducerIdsRequest)
+        case ApiKeys.DESCRIBE_QUORUM => forwardToControllerOrFail(request)
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaApis.scala:230]: Hm, so we are _only_ allowing forwarding with this `forwardToControllerOrFail` since this RPC did not exist before? Are there other recently added controller-only RPCs that need this treatment?

**Review by mumrah — APPROVED** [review]
Thanks @hachikuji, LGTM. Two non-blocking questions inline

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 5 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala:230]: Yeah, because DescribeQuorum does not make sense for zk clusters. As far as I know, this is the only case of a client api which is only exposed under kraft. However, we do have some internal controller apis that are only used by kraft (e.g. for broker registration and heartbeating).

File: core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala (hunk 4 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:36]: Yeah, good idea. Added in the latest commit.
